### PR TITLE
RPE-210: Prepare metrics feature

### PR DESCRIPTION
**Pull request by doncem**
### Change description ###

Minor code review to ease up metrics integration:

- letter validation
- ftp uploader

Additional code review:

- line breaks
- gradle dependency order (OCD or w/e)

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/slc/services/steps/maptoletter/LetterMapper.java [hunk 1]
```diff
@@ -15,16 +19,19 @@
 public class LetterMapper {
 
     private static final ObjectMapper objectMapper = new ObjectMapper();
+    private static final Validator validator = Validation.buildDefaultValidatorFactory().getValidator();
 
     public Letter from(IMessage msg) {
         try {
             Letter letter = objectMapper.readValue(msg.getBody(), Letter.class);
-            if (letter != null) {
+            Set<ConstraintViolation<Letter>> violations = validator.validate(letter);
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/reform/slc/services/steps/maptoletter/LetterMapper.java]: why do we need to validate this?Letter is validated before persisting into queue.

File: src/main/java/uk/gov/hmcts/reform/slc/model/Letter.java [hunk 2]
```diff
@@ -3,12 +3,17 @@
...
 public class Letter {
 
+    @NotNull
     public final String template;
+    @NotNull
     public final Map<String, Object> values;
+    @NotNull
     public final String type;
+    @NotNull
     public final String service;
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/slc/model/Letter.java]: is this validation ever used? What if it fails?

File: src/main/java/uk/gov/hmcts/reform/slc/services/steps/maptoletter/LetterMapper.java [hunk 3]
```diff
@@ -15,16 +19,19 @@
 public class LetterMapper {
 
     private static final ObjectMapper objectMapper = new ObjectMapper();
+    private static final Validator validator = Validation.buildDefaultValidatorFactory().getValidator();
 
     public Letter from(IMessage msg) {
         try {
             Letter letter = objectMapper.readValue(msg.getBody(), Letter.class);
-            if (letter != null) {
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/slc/services/steps/maptoletter/LetterMapper.java]: is this check no longer needed?

File: src/main/java/uk/gov/hmcts/reform/slc/model/Letter.java [hunk 4]
```diff
@@ -3,12 +3,17 @@
...
 
 public class Letter {
 
+    @NotNull
     public final String template;
+    @NotNull
     public final Map<String, Object> values;
+    @NotNull
     public final String type;
+    @NotNull
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/slc/model/Letter.java]: what about empty string?

File: src/main/java/uk/gov/hmcts/reform/slc/services/steps/maptoletter/LetterMapper.java (hunk 1 above)
Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/slc/services/steps/maptoletter/LetterMapper.java]: There is no such thing as `letter == null` when you unmarshall the letter. It's either exception throw or letter object with null fields inside. So actually we need to test all the fields of null of letter object and not the object being null itself.

Hope that makes sense.

Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/slc/services/steps/maptoletter/LetterMapper.java]: there is, I just checked:
```
Letter letter = new ObjectMapper().readValue("null", Letter.class);
assertThat(letter).isNull();
```

Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/slc/services/steps/maptoletter/LetterMapper.java]: 😭

File: src/main/java/uk/gov/hmcts/reform/slc/services/steps/maptoletter/LetterMapper.java [hunk 5]
```diff
@@ -15,18 +19,28 @@
...
+            return validate(objectMapper.readValue(msg.getBody(), Letter.class));
+        } catch (IOException exc) {
+            throw new InvalidMessageException("Unable to deserialize message " + msg.getMessageId(), exc);
+        }
+    }
+
+    private Letter validate(Letter letter) throws InvalidMessageException {
+        if (letter != null) {
+            Set<ConstraintViolation<Letter>> violations = validator.validate(letter);
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/reform/slc/services/steps/maptoletter/LetterMapper.java]: I am not sure why you need to validate the letter.Are you expecting message to be modified once it is persisted into queue?

Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/slc/services/steps/maptoletter/LetterMapper.java]: It can be once it is resubmitted from the dead letter queue

Comment (nitinprabhu) [reply src/main/java/uk/gov/hmcts/reform/slc/services/steps/maptoletter/LetterMapper.java]: Also code could be rewritten as below instead of if else if else
```
private Letter validate(Letter letter) throws InvalidMessageException {
		if(letter==null){
			throw new InvalidMessageException("Invalid message body");
		}

		Set<ConstraintViolation<Letter>> violations = validator.validate(letter);

		if(!violations.isEmpty()){
			throw new InvalidMessageException("Invalid message body");
         }

        return letter;
 }

```

Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/slc/services/steps/maptoletter/LetterMapper.java]: looks like we need pmd rule enabled

File: src/main/java/uk/gov/hmcts/reform/slc/services/steps/maptoletter/LetterMapper.java [hunk 6]
```diff
@@ -15,18 +19,28 @@
...
-
+            return validate(objectMapper.readValue(msg.getBody(), Letter.class));
         } catch (IOException exc) {
             throw new InvalidMessageException("Unable to deserialize message " + msg.getMessageId(), exc);
         }
     }
+
+    private Letter validate(Letter letter) throws InvalidMessageException {
+        if (letter == null) {
+            throw new InvalidMessageException("Invalid message body");
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/slc/services/steps/maptoletter/LetterMapper.java]: I think "Empty message ..." was better

File: src/test/java/uk/gov/hmcts/reform/slc/services/steps/maptoletter/LetterMapperTest.java [hunk 7]
```diff
@@ -42,15 +43,51 @@ public void should_return_letter_if_message_contains_valid_json() {
...
             .willReturn("{\"a\" : \"b\"}".getBytes());
 
         assertThatThrownBy(() -> letterMapper.from(message))
-            .isInstanceOf(InvalidMessageException.class);
+            .isInstanceOf(InvalidMessageException.class)
+            .hasMessageStartingWith("Unable to deserialize message");
+    }
+
+    @Test
+    public void should_throw_an_exception_if_message_contains_null_fields() {
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/slc/services/steps/maptoletter/LetterMapperTest.java]: `{}` contains null fields?

File: src/test/java/uk/gov/hmcts/reform/slc/services/steps/maptoletter/LetterMapperTest.java [hunk 8]
```diff
@@ -42,15 +43,51 @@ public void should_return_letter_if_message_contains_valid_json() {
...
+            ).getBytes()
+        );
+
+        assertThatThrownBy(() -> letterMapper.from(message))
+            .isInstanceOf(InvalidMessageException.class)
+            .hasMessageStartingWith("Invalid message body");
+    }
+
+    @Test
+    public void should_throw_aan_exception_if_message_contains_null_fields() {
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/slc/services/steps/maptoletter/LetterMapperTest.java]: typo in 'aan'

File: src/test/java/uk/gov/hmcts/reform/slc/services/steps/maptoletter/LetterMapperTest.java (hunk 7 above)
Comment (doncem) [reply src/test/java/uk/gov/hmcts/reform/slc/services/steps/maptoletter/LetterMapperTest.java]: All fields unmarshall as null. Don't want to prove the case with `4!` attempts for 3 scenarios just to make point code branch works

File: src/test/java/uk/gov/hmcts/reform/slc/services/steps/maptoletter/LetterMapperTest.java (hunk 8 above)
Comment (doncem) [reply src/test/java/uk/gov/hmcts/reform/slc/services/steps/maptoletter/LetterMapperTest.java]: ah snap, that was quick copy paste

File: src/test/java/uk/gov/hmcts/reform/slc/services/steps/maptoletter/LetterMapperTest.java (hunk 7 above)
Comment (rkondratowicz) [reply src/test/java/uk/gov/hmcts/reform/slc/services/steps/maptoletter/LetterMapperTest.java]: I mean that method description doesn't match value used in test

File: src/test/java/uk/gov/hmcts/reform/slc/services/steps/maptoletter/LetterMapperTest.java [hunk 9]
```diff
@@ -42,15 +43,51 @@ public void should_return_letter_if_message_contains_valid_json() {
...
             .willReturn("{\"a\" : \"b\"}".getBytes());
 
         assertThatThrownBy(() -> letterMapper.from(message))
-            .isInstanceOf(InvalidMessageException.class);
+            .isInstanceOf(InvalidMessageException.class)
+            .hasMessageStartingWith("Unable to deserialize message");
+    }
+
+    @Test
+    public void should_throw_an_exception_if_message_contains_empty_body_representing_all_fields_as_null() {
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/slc/services/steps/maptoletter/LetterMapperTest.java]: 'empty_body_representing_all_fields_as_null' or 'empty_json'? (it's `{}`)

**Review by rkondratowicz — APPROVED** [review]
